### PR TITLE
[11.0][MIG] account_early_payment_discount

### PR DESCRIPTION
**Pull request by asaunier**
Migration of a module proposed in 9.0, see https://github.com/OCA/account-payment/pull/85

### CODE REVIEW

**Comment by asaunier** [comment]
I don't understand why Travis complains:
```
2018-04-13 15:26:47,957 5038 INFO openerp_test odoo.addons.account_early_payment_discount.tests.test_payment_term: test_constraint__failed (odoo.addons.account_early_payment_discount.tests.test_payment_term.TestPaymentTerm)
2018-04-13 15:26:47,979 5038 ERROR openerp_test odoo.sql_db: bad query: b'INSERT INTO "account_payment_term" ("id", "early_payment_discount", "name", "company_id", "sequence", "active", "create_uid", "write_uid", "create_date", "write_date") VALUES(nextval(\'account_payment_term_id_seq\'), true, \'Unittest payment term\', 1, 10, true, 1, 1, (now() at time zone \'UTC\'), (now() at time zone \'UTC\')) RETURNING id'
ERROR: new row for relation "account_payment_term" violates check constraint "account_payment_term_early_payment_discount"
DETAIL:  Failing row contains (6, Unittest payment term, t, null, 1, 10, 1, 2018-04-13 15:26:47.958971, 1, 2018-04-13 15:26:47.958971, t, null, null, null).
```
Test ``test_constraint__failed`` indeed performs an illegal action by creating a payment term with no ``epd_nb_days`` and ``epd_discount``, breaking the [early_payment_discount constraint](https://github.com/asaunier/account-payment/blob/26357314b607ac512719bec2dd8ad856a60ad53e/account_early_payment_discount/models/account_payment_term.py#L29-L32).

But that's the goal of the test to make sure an integrity error is raised in such a case and a special ``assertRaises`` assertion is actually used:
https://github.com/asaunier/account-payment/blob/26357314b607ac512719bec2dd8ad856a60ad53e/account_early_payment_discount/tests/test_payment_term.py#L12

**Comment by asaunier** [comment]
8e1726d fixes the problem reported above when running the tests.

It seems the ``maintainer-quality-tools`` think tests have failed when an ``ERROR`` keyword is found in the logs. In our case the ``ERROR`` is an expected integrity error when testing an illegal action that breaks a constraint (the test actually passes).

Before this change in Odoo 9 months ago https://github.com/odoo/odoo/commit/c4e9b854ff22af31bbbd88d1414e77d27eaed7ea (see odoo/sql_db.py) such logs were considered as ``INFO``, not ``ERROR``.

8e1726d then prevents the error from being logged using cursor's ``_default_log_exceptions`` property, introduced by https://github.com/odoo/odoo/commit/e85f246469d6dbf5c9f801e20babd7d13b2c05f9

I have added an issue in ``maintainer-quality-tools`` repo at https://github.com/OCA/maintainer-quality-tools/issues/542

This PR is ready for review!

File: account_early_payment_discount/tests/test_payment_term.py
Comment (simahawk) [inline account_early_payment_discount/tests/test_payment_term.py]: remove this and decorate the method w/ `@mute_logger('odoo.sql_db')`

File: account_early_payment_discount/README.rst
Comment (simahawk) [inline account_early_payment_discount/README.rst]: use png

**Review by simahawk — CHANGES_REQUESTED** [review]

**Review by tschanzt — APPROVED** [review]

**Comment by asaunier** [comment]
@simahawk @tschanzt Thanks for the feedback and tips!
I have updated/rebased the PR to use ``@mute_logger('odoo.sql_db')`` instead and the PNG file for the AGPL icon.

**Review by yvaucher — APPROVED** [review]

**Comment by feketemihai** [comment]
@asaunier Did you check the PR #174 , it basically adds the same concept, but in more complex, handle the early payment writeoff, with the module that you propose you only have the info on the payment term, so we will still need a module for writeoff the discount. I suggest to port that one since is handling the same requirement but in a more complex workflow.

**Review by simahawk — APPROVED** [review]

File: account_early_payment_discount/models/account_payment_term.py
Comment (grindtildeath) [inline account_early_payment_discount/models/account_payment_term.py]: s/should/must as it's an imperative.

File: account_early_payment_discount/models/account_payment_term.py
Comment (grindtildeath) [inline account_early_payment_discount/models/account_payment_term.py]: It seems you can remove this one.

File: account_early_payment_discount/tests/test_payment_term.py
Comment (grindtildeath) [inline account_early_payment_discount/tests/test_payment_term.py:28]: asserting something would be nice.

File: account_early_payment_discount/README.rst
Comment (grindtildeath) [inline account_early_payment_discount/README.rst]: The module will be called 'account_reconcile_rule_early_payment_discount' (to be confirmed by @fclementic2c ) but its repo is now OCA/account-reconcile.

File: account_early_payment_discount/views/account_payment_term.xml
Comment (grindtildeath) [inline account_early_payment_discount/views/account_payment_term.xml]: Why don't you inherit `account.view_payment_term_tree` ?

File: account_early_payment_discount/views/account_payment_term.xml
Comment (grindtildeath) [inline account_early_payment_discount/views/account_payment_term.xml]: A name is missing for this view

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Going to try it out migrating account_operation_rule_early_payment_discount ;)

File: account_early_payment_discount/tests/test_payment_term.py
Comment (asaunier) [reply account_early_payment_discount/tests/test_payment_term.py:28]: This test simply checks that it is possible to create a payment term if the mandatory attributes 'epd_nb_days' and 'epd_discount' are provided. Which is not the case in the previous test where those attributes are omitted, causing an integrity error.
The assert is then implicit here.

File: account_early_payment_discount/views/account_payment_term.xml
Comment (asaunier) [reply account_early_payment_discount/views/account_payment_term.xml]: I guess you mean https://github.com/odoo/odoo/blob/dccef8e41d45dd5150e4112aa3a2b19ba698dc62/addons/account/views/account_view.xml#L1631-L1640

I don't have the answer to your question: I have only migrated the module, not authored it :)
Then I don't know if there is a reason for replacing the view here.

File: account_early_payment_discount/views/account_payment_term.xml
Comment (asaunier) [reply account_early_payment_discount/views/account_payment_term.xml]: Is it really required?

**Comment by asaunier** [comment]
@grindtildeath Thanks for reviewing!
I have just pushed a fixup taking into account some of your suggestions.

File: account_early_payment_discount/tests/test_payment_term.py
Comment (grindtildeath) [reply account_early_payment_discount/tests/test_payment_term.py:28]: IMO you should then assert the field on the created record has the value you provided. If someone breaks the creation doing any change the test will still pass. 
Moreover "Explicit is better than implicit" PEP 20 :wink:

File: account_early_payment_discount/views/account_payment_term.xml
Comment (grindtildeath) [reply account_early_payment_discount/views/account_payment_term.xml]: Actually the module removes the `sequence` field which could be of some use. I don't think it's with intent ?
The reason looks simply to be, that this view did not exist prior to v11, but as it now does, IMO we should respect the inheritance.

File: account_early_payment_discount/views/account_payment_term.xml
Comment (grindtildeath) [reply account_early_payment_discount/views/account_payment_term.xml]: No, it's not, but it makes no harm and it's better to have a name when you have to select a record for any field on the GUI.

**Comment by asaunier** [comment]
@grindtildeath OK with your arguments. I have made a fixup commit to take them into account.

_Edit: squashed the fixup commits_

**Review by grindtildeath — APPROVED** [review]
Thank you @asaunier :+1:

**Comment by asaunier** [comment]
Fixup commits have been squashed. This PR can be merged if no other objection is raised. Thanks :)

**Comment by rafaelbn** [comment]
Hello! Nobody answer to https://github.com/OCA/account-payment/pull/184#issuecomment-383139170 of @feketemihai about https://github.com/OCA/account-payment/pull/174 :thinking: any case thank! I will try it! :smile:

**Pull request merged**
